File: amazon-clone/src/store/cartStore.ts
import { create } from "zustand";
import { persist } from "zustand/middleware";
import { Product, CartItem } from "@/lib/types";
import { SHIPPING_THRESHOLD, SHIPPING_COST, TAX_RATE } from "@/lib/constants";

interface CartState {
  items: CartItem[];
  savedItems: CartItem[];
  addItem: (product: Product, quantity?: number) => void;
  removeItem: (productId: string) => void;
  updateQuantity: (productId: string, quantity: number) => void;
  clearCart: () => void;
  saveForLater: (productId: string) => void;
  moveToCart: (productId: string) => void;
  removeSavedItem: (productId: string) => void;
  totalItems: () => number;
  subtotal: () => number;
  tax: () => number;
  shipping: () => number;
  total: () => number;
}

export const useCartStore = create<CartState>()(
  persist(
    (set, get) => ({
      items: [],
      savedItems: [],

      addItem: (product: Product, quantity = 1) => {
        set((state) => {
          const existingItem = state.items.find(
            (item) => item.product.id === product.id
          );
          if (existingItem) {
            return {
              items: state.items.map((item) =>
                item.product.id === product.id
                  ? { ...item, quantity: item.quantity + quantity }
                  : item
              ),
            };
          }
          return { items: [...state.items, { product, quantity }] };
        });
      },

      removeItem: (productId: string) => {
        set((state) => ({
          items: state.items.filter((item) => item.product.id !== productId),
        }));
      },

      updateQuantity: (productId: string, quantity: number) => {
        if (quantity < 1) return;
        set((state) => ({
          items: state.items.map((item) =>
            item.product.id === productId ? { ...item, quantity } : item
          ),
        }));
      },

      clearCart: () => set({ items: [] }),

      saveForLater: (productId: string) => {
        const item = get().items.find((i) => i.product.id === productId);
        if (!item) return;
        set((state) => ({
          items: state.items.filter((i) => i.product.id !== productId),
          savedItems: [...state.savedItems, { ...item, quantity: 1 }],
        }));
      },

      moveToCart: (productId: string) => {
        const item = get().savedItems.find((i) => i.product.id === productId);
        if (!item) return;
        set((state) => {
          const existingItem = state.items.find(
            (i) => i.product.id === productId
          );
          if (existingItem) {
            return {
              savedItems: state.savedItems.filter(
                (i) => i.product.id !== productId
              ),
              items: state.items.map((i) =>
                i.product.id === productId
                  ? { ...i, quantity: i.quantity + 1 }
                  : i
              ),
            };
          }
          return {
            savedItems: state.savedItems.filter(
              (i) => i.product.id !== productId
            ),
            items: [...state.items, { ...item, quantity: 1 }],
          };
        });
      },

      removeSavedItem: (productId: string) => {
        set((state) => ({
          savedItems: state.savedItems.filter(
            (i) => i.product.id !== productId
          ),
        }));
      },

      totalItems: () => {
        return get().items.reduce((total, item) => total + item.quantity, 0);
      },

      subtotal: () => {
        return get().items.reduce(
          (total, item) => total + item.product.price * item.quantity,
          0
        );
      },

      tax: () => {
        return get().subtotal() * TAX_RATE;
      },

      shipping: () => {
        const subtotal = get().subtotal();
        if (subtotal === 0) return 0;
        return subtotal >= SHIPPING_THRESHOLD ? 0 : SHIPPING_COST;
      },

      total: () => {
        return get().subtotal() + get().tax() + get().shipping();
      },
    }),
    {
      name: "novamart-cart",
    }
  )
);

File: ios-amazon-clone/NovaMart/Stores/CartStore.swift
import Foundation
import Observation

@Observable
final class CartStore {
    var items: [CartItem] = []
    var savedItems: [CartItem] = []

    init() {
        self.items = PersistenceService.load([CartItem].self, forKey: PersistenceService.Keys.cartItems) ?? []
        self.savedItems = PersistenceService.load([CartItem].self, forKey: PersistenceService.Keys.savedItems) ?? []
    }

    // MARK: - Computed Properties

    var totalItems: Int {
        items.reduce(0) { $0 + $1.quantity }
    }

    var subtotal: Double {
        items.reduce(0) { $0 + $1.product.price * Double($1.quantity) }
    }

    var tax: Double {
        subtotal * Constants.taxRate
    }

    var shipping: Double {
        guard subtotal > 0 else { return 0 }
        return subtotal >= Constants.shippingThreshold ? 0 : Constants.shippingCost
    }

    var total: Double {
        subtotal + tax + shipping
    }

    // MARK: - Actions

    func addItem(_ product: Product, quantity: Int = 1) {
        if let index = items.firstIndex(where: { $0.product.id == product.id }) {
            items[index].quantity += quantity
        } else {
            items.append(CartItem(product: product, quantity: quantity))
        }
        persistCart()
    }

    func removeItem(_ productId: String) {
        items.removeAll { $0.product.id == productId }
        persistCart()
    }

    func updateQuantity(_ productId: String, quantity: Int) {
        guard quantity >= 1 else { return }
        if let index = items.firstIndex(where: { $0.product.id == productId }) {
            items[index].quantity = quantity
        }
        persistCart()
    }

    func clearCart() {
        items.removeAll()
        persistCart()
    }

    func saveForLater(_ productId: String) {
        guard let item = items.first(where: { $0.product.id == productId }) else { return }
        items.removeAll { $0.product.id == productId }
        if !savedItems.contains(where: { $0.product.id == productId }) {
            savedItems.append(CartItem(product: item.product, quantity: 1))
        }
        persistCart()
        persistSaved()
    }

    func moveToCart(_ productId: String) {
        guard let item = savedItems.first(where: { $0.product.id == productId }) else { return }
        savedItems.removeAll { $0.product.id == productId }
        if let index = items.firstIndex(where: { $0.product.id == productId }) {
            items[index].quantity += 1
        } else {
            items.append(CartItem(product: item.product, quantity: 1))
        }
        persistCart()
        persistSaved()
    }

    func removeSavedItem(_ productId: String) {
        savedItems.removeAll { $0.product.id == productId }
        persistSaved()
    }

    // MARK: - Persistence

    private func persistCart() {
        PersistenceService.save(items, forKey: PersistenceService.Keys.cartItems)
    }

    private func persistSaved() {
        PersistenceService.save(savedItems, forKey: PersistenceService.Keys.savedItems)
    }
}
